sympy/solvers/solveset.py

```diff
@@ -566,13 +572,16 @@ def _solve_trig1(f, symbol, domain):
     if isinstance(solns, FiniteSet):
         if any(isinstance(s, RootOf) for s in solns):
             raise NotImplementedError
-        result = Union(*[invert_complex(exp(I*symbol), s, symbol)[1]
-                       for s in solns])
-        return Intersection(result, domain)
+        result = Union(*[inverter(cov, s, symbol)[1] for s in solns])
+        # avoid spurious intersections with C in solution set
+        if domain is S.Complexes:
+            return result
+        else:
+            return Intersection(result, domain)
```

Comment:
is there nothing higher up to handle this? This doesn't seem like the right place to me. Many operations in solveset can lead to an intersection like this so simplifying those could be handled at the end. Maybe `Intersection` can be improved to handle intersections with `Complexes`.

Reply:
Yes, it may be a bit hackish to have this here, but I'd say it's not as bad as it may seem:

In general the various solveset helpers mostly return `FiniteSet`s for which intersection with `Complexes` is handled just fine, but for trig equations there are often (infinite) `ImageSet`s that (currently) don't mix well with intersections. In that sense this "intersection-removal" is somewhat specific for trig equations. Or at least here the spurious intersections are quite prominent and annoying, which prompted me to insert this workaround.
```
>>> solveset(sinh(x), x, Complexes)
{2⋅n⋅ⅈ⋅π | n ∊ ℤ} ∪ {ⅈ⋅(2⋅n⋅π + π) | n ∊ ℤ}
>>> _.intersect(Complexes)
(ℂ ∩ {2⋅n⋅ⅈ⋅π | n ∊ ℤ}) ∪ (ℂ ∩ {ⅈ⋅(2⋅n⋅π + π) | n ∊ ℤ})
```

Moving this intersection-removal too high up the call chain could also be problematic, since the origin of the spurious intersection is less well controlled.

I do concur though that it'd be best if the intersection in general could be improved, especially for "trivial" examples as shown above. (There's a recent issue for this, as I see now: #17996).